deepdiff/model.py

```diff
@@ -0,0 +1,493 @@
+# -*- coding: utf-8 -*-
+
+from .helper import items
+from .helper import is_string
+from .helper import RemapDict
+from abc import ABCMeta, abstractmethod
+from ast import literal_eval
+from copy import copy
+
+
+class ResultDict(RemapDict):
+    def cleanup(self):
+        """
+        Remove empty keys from this object. Should always be called after the result is final.
+        :return:
+        """
+        empty_keys = [k for k, v in getattr(self, items)() if not v]
+
+        for k in empty_keys:
+            del self[k]
+
+
+class RefStyleResultDict(ResultDict):
+    def __init__(self):
+        # TODO: centralize keys
+        self.update({"type_changes": set(),
+                     "dictionary_item_added": set(), "dictionary_item_removed": set(),
+                     "values_changed": set(), "unprocessed": set(),
+                     "iterable_item_added": set(), "iterable_item_removed": set(),
+                     "attribute_added": set(), "attribute_removed": set(),
+                     "set_item_removed": set(), "set_item_added": set(),
+                     "repetition_change": set()})
+
+
+class TextStyleResultDict(ResultDict):
+    def __init__(self, verbose_level = 1, ref_results = None):
+        self.verbose_level = verbose_level
+
+        # TODO: centralize keys
+        self.update({"type_changes": {}, "dictionary_item_added": self.__set_or_dict(),
+                     "dictionary_item_removed": self.__set_or_dict(),
+                     "values_changed": {}, "unprocessed": [], "iterable_item_added": {}, "iterable_item_removed": {},
+                     "attribute_added": self.__set_or_dict(), "attribute_removed": self.__set_or_dict(),
+                     "set_item_removed": set([]),
+                     "set_item_added": set([]), "repetition_change": {}})
+
+        if ref_results:
+            self.from_ref_results(ref_results)
+
+    def __set_or_dict(self):
+        return {} if self.verbose_level >= 2 else set()
+        #return dict()
+
+    def from_ref_results(self, ref):
+        """
+        Populate this object by parsing an existing reference-style result dictionary.
+        :param ref: A RefStyleResultDict
+        :return:
+        """
+        self._from_ref_type_changes(ref)
+        self._from_ref_default(ref, 'dictionary_item_added')
+        self._from_ref_default(ref, 'dictionary_item_removed')
+        self._from_ref_value_changed(ref)
+        self._from_ref_unprocessed(ref)
+        self._from_ref_default(ref, 'iterable_item_added')
+        self._from_ref_default(ref, 'iterable_item_removed')
+        self._from_ref_default(ref, 'attribute_added')
+        self._from_ref_default(ref, 'attribute_removed')
+        self._from_ref_set_item_removed(ref)
+        self._from_ref_set_item_added(ref)
+        self._from_ref_repetition_change(ref)
+
+    def _from_ref_default(self, ref, type):
+        if type in ref:
```

Comment:
It is generally recommended not to use builtin names as variable names. Maybe `type` should be called something else like `ref_item` or something?